api/test/wfl/integration/source_test.clj

```diff
@@ -37,18 +47,94 @@
 
 (defn ^:private create-tdr-source []
   (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
-    (->> {:name           "Terra DataRepo",
-          :dataset        "this"
-          :table          "is"
-          :column         "fun"
-          :skipValidation true}
+    (->> {:column         "a-column"
+          :dataset        {:name "a-dataset"}
+          :id             "an-id"
+          :name           "Terra DataRepo"
+          :skipValidation true
+          :table          "a-table"
+          :type           "a-type"}
          (source/create-source! tx (rand-int 1000000))
          (zipmap [:source_type :source_items])
          (source/load-source! tx))))
 
 (defn ^:private reload-source [tx {:keys [type id] :as _source}]
   (source/load-source! tx {:source_type type :source_items (str id)}))
 
+;; A stable vector of TDR row IDs.
+;;
+(defonce all-rows
+  (delay (vec (repeatedly mock-new-rows-size #(str (UUID/randomUUID))))))
+
+(def fibonacci
+  "The Fibonacci sequence."
+  (lazy-cat [0 1] (map + fibonacci (rest fibonacci))))
+
+(defn ^:private datarepo-query-table-between-all
+  "Mock datarepo/query-table-between to find all rows."
+  [_dataset _table _between interval columns]
+  (is (every? parse-timestamp interval))
+  (letfn [(field [column] {:mode "NULLABLE" :name column :type "STRING"})
+          (fields [columns] (mapv field columns))]
+    (-> {:jobComplete true
+         :kind "bigquery#queryResponse"
+         :rows [@all-rows]
+         :totalRows (str (count @all-rows))}
+        (assoc-in [:schema :fields] (fields columns)))))
+
+(defn ^:private datarepo-query-table-between-miss
+  "Mock datarepo/query-table-between to miss some rows."
+  [dataset table between interval columns]
+  (let [miss (take-while (partial > mock-new-rows-size) fibonacci)
```

Comment:
Why fibonacci? Yes it's a nice implementation but not worth the conceptual burden. Can you take n instead?

Reply:
> Why fibonacci? Yes it's a nice implementation but not worth the conceptual burden. Can you take n instead?

Fibonacci felt like the sort of distribution we'd actually see:
several thick off the front with several stragglers following.
Didn't want to assume that missing rows were all at the front
of the interval.

Reply:
Maybe the right thing to do is create a dataset specifically for this test. This smells like square pegs and round holes.

Reply:
Looking at the test, it's not immediately apparent what behaviour it's supposed to lock down.

Reply:
Not sure what you mean.  Should the test just drop some rows from the start?
The test mocks the TDR and the dataset (mostly) specifically for this test.

Reply:
I mean my main issue is that I don't understand what the fibonacci numbers have to do with missing a row when polling tdr.

I didn't remember if this test mocked TDR, thank you for clarifying that. 

It would be simpler if you returned one less that all the rows you're expecting the first time you poll, then allow all the rows to be found next time, with the assertion that the lower bound of the time interval is far back enough.

Reply:
The test is more comprehensive than that now.
It removes a set of rows scattered in the interval
from the results of the first query, then verifies
that only those rows show up in the cleanup round.
I can try to make that more obvious, but it sounds
like the test does not mock appropriate conditions,
or that the solution is wrong.  Maybe we should talk
after I address the bugs you found?

Reply:
> It removes a set of rows scattered in the interval

Why does it need to do that? what's wrong with removing all but the last?

Reply:
All but the last row?

Reply:
Let's say you're mocking a BiqQuery query that's supposed to return a list 5 of rows that workflow-launcher will then snapshot. Instead of returning all 5 rows for the time interval specified, make the mock return all but the last row. Then in your test, assert that the next time wfl queries for the new rows (or when it next tries to find rows it had previously missed), the missed row is in the query interval.

Reply:
As you wish.

Reply:
I'd prefer tests to be simple and not have additional logic that serves as a red herring.

Reply:
Not a fan of herring myself.